core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1004,19 +999,25 @@ private void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, Ex
                     // remote log segments won't be removed. The `isRemoteSegmentWithinLeaderEpoch` validates whether
                     // the epochs present in the segment lies in the checkpoint file. It will always return false
                     // since the checkpoint file was already truncated.
-                    boolean isSegmentDeleted = remoteLogRetentionHandler.deleteLogStartOffsetBreachedSegments(
+                    boolean shouldDeleteSegment = remoteLogRetentionHandler.deleteLogStartOffsetBreachedSegments(
                             metadata, logStartOffset, epochWithOffsets);
+                    if (shouldDeleteSegment) {
+                        segmentsToDelete.add(metadata);
+                    }
                     boolean isValidSegment = false;
-                    if (!isSegmentDeleted) {
+                    if (!shouldDeleteSegment) {
                         // check whether the segment contains the required epoch range with in the current leader epoch lineage.
                         isValidSegment = isRemoteSegmentWithinLeaderEpochs(metadata, logEndOffset, epochWithOffsets);
                         if (isValidSegment) {
-                            isSegmentDeleted =
+                            shouldDeleteSegment =
                                     remoteLogRetentionHandler.deleteRetentionTimeBreachedSegments(metadata) ||
                                             remoteLogRetentionHandler.deleteRetentionSizeBreachedSegments(metadata);
+                            if (shouldDeleteSegment) {
+                                segmentsToDelete.add(metadata);
+                            }
```

Comment:
nit: These if block and L1004-1006 could be put before `canProcess = isSegmentDeleted || !isValidSegment;`. I.e.
```
boolean shouldDeleteSegment = remoteLogRetentionHandler.deleteLogStartOffsetBreachedSegments(
                            metadata, logStartOffset, epochWithOffsets);
boolean isValidSegment = false;

if (!shouldDeleteSegment) {
    ...
  if (isValidSegment) {
      shouldDeleteSegment = remoteLogRetentionHandler.deleteRetentionTimeBreachedSegments(metadata) || remoteLogRetentionHandler.deleteRetentionSizeBreachedSegments(metadata);
  }
}
if (shouldDeleteSegment) {
   segmentsToDelete.add(metadata);
}

canProcess = isSegmentDeleted || !isValidSegment;
```

Reply:
Thanks for the spot! Hopefully addressed in the next commit